fast path for sympy gcd in floordiv

avikchaudhuri opened the pull request:
Summary:
Re-implementation of https://github.com/pytorch/pytorch/pull/134150, which was reverted because of some internal tests hanging (case B). The original motivation was to get some other internal test unstuck (case A).

The root cause is that sympy.gcd is both very clever as well as can blow up in some cases. This PR introduces a fast path with an appropriate fallback to sympy.gcd that ensures that both cases A and B go through.

Test Plan:
See the included test for specific examples.
Also https://fb.workplace.com/groups/1075192433118967/posts/1491493248155548/?comment_id=1491938994777640&reply_comment_id=1492622821375924

Differential Revision: D62043315


cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

ezyang commented:
CI looks like some sort of bug, holler if you want review now anyway

avikchaudhuri commented:
> CI looks like some sort of bug, holler if you want review now anyway

Should be good to review now. A `map` object was being iterated on multiple times, wrapping it with `list` fixed the bug.

ezyang commented on torch/utils/_sympy/functions.py:
```diff
@@ -202,7 +242,9 @@ def eval(cls, base, divisor):
-            gcd = sympy.gcd(base, divisor)
+            gcd = simple_floordiv_gcd(base, divisor)
+            if equal_valued(gcd, 1) and isinstance(divisor, sympy.Add):
+                gcd = sympy.gcd(base, divisor)
```
Oh! Is this one load bearing? I guess it is if the divisor got expanded.

avikchaudhuri replied:
Yeah, unfortunately I ran into the `(x*y + x + y + 1) // (x + 1)` during internal testing and I don't want to invent a fancy new gcd algo lol. (In this case I could subtract and factorize, but general gcd probably would need repeated subtraction Euclid style.)

I also tried a version that uses `sympy.factor` but that loops on some inputs as well, so as bad as `sympy.gcd`.

ezyang commented on torch/utils/_sympy/functions.py:
```diff
@@ -95,6 +95,46 @@ def fuzzy_eq(x, y):
+
+    def integer_factor(expr):
+        integer_factors = map(integer_coefficient, sympy.Add.make_args(expr))
+        return functools.reduce(math.gcd, integer_factors)
```
Uhhh doesn't math.gcd support arbitrarily many arguments lol

ezyang replied:
Oh, Python 3.9 only. Well, consider doing a version test...

avikchaudhuri replied:
So this improved in 3.9 to support arbitrarily many args, but I've been bitten by this once before so kept the binary args.

ezyang commented:
With this newest implementation, what's the performance differential? Also, we've been adding compile tests to benchmarks/dynamo/pr_time_benchmarks/benchmarks, consider adding a representative example there so we can track its performance over time.

ezyang reviewed (approved):
looks functionally correct

ezyang commented:
BTW, you might consider killswitching this one, if you're still worried about unintended consequences. Not sure what your current risk assessment is, I think your intuition is fine here.

avikchaudhuri commented:
> BTW, you might consider killswitching this one, if you're still worried about unintended consequences. Not sure what your current risk assessment is, I think your intuition is fine here.

I'm less worried now because of the fallback.

avikchaudhuri committed:
fast path for sympy gcd in floordiv (#134880)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/134880

Re-implementation of https://github.com/pytorch/pytorch/pull/134150, which was reverted because of some internal tests hanging (case B). The original motivation was to get some other internal test unstuck (case A).

The root cause is that sympy.gcd is both very clever as well as can blow up in some cases. This PR introduces a fast path with an appropriate fallback to sympy.gcd that ensures that both cases A and B go through.

Test Plan:
See the included test for specific examples.
Also https://fb.workplace.com/groups/1075192433118967/posts/1491493248155548/?comment_id=1491938994777640&reply_comment_id=1492622821375924

Reviewed By: angelayi

Differential Revision: D62043315
